File: hold-it-ios/ViewModels/AppState.swift
//
//  AppState.swift
//  hold-it-ios
//

import SwiftUI
import SwiftData
import WidgetKit

// MARK: - 货币配置

/// 支持的货币列表
enum SupportedCurrency: String, CaseIterable, Identifiable {
    case auto   = "auto"
    case cny    = "CNY"
    case usd    = "USD"
    case eur    = "EUR"
    case jpy    = "JPY"
    case gbp    = "GBP"
    case hkd    = "HKD"
    case twd    = "TWD"
    case krw    = "KRW"

    var id: String { rawValue }

    /// 显示名称（本地化）
    var displayName: String {
        switch self {
        case .auto: return String(localized: "跟随系统")
        case .cny:  return String(localized: "人民币 (¥)")
        case .usd:  return String(localized: "美元 ($)")
        case .eur:  return String(localized: "欧元 (€)")
        case .jpy:  return String(localized: "日元 (¥)")
        case .gbp:  return String(localized: "英镑 (£)")
        case .hkd:  return String(localized: "港元 (HK$)")
        case .twd:  return String(localized: "台币 (NT$)")
        case .krw:  return String(localized: "韩元 (₩)")
        }
    }

    /// 对应的货币符号
    var symbol: String {
        switch self {
        case .auto: return SupportedCurrency.systemSymbol
        case .cny:  return "¥"
        case .usd:  return "$"
        case .eur:  return "€"
        case .jpy:  return "¥"
        case .gbp:  return "£"
        case .hkd:  return "HK$"
        case .twd:  return "NT$"
        case .krw:  return "₩"
        }
    }

    /// 读取系统当前 Locale 对应的货币符号
    static var systemSymbol: String {
        guard let code = Locale.current.currency?.identifier else { return "¥" }
        // 用系统 Locale formatter 拿到简短符号
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencyCode = code
        formatter.locale = Locale.current
        return formatter.currencySymbol ?? "¥"
    }
}

@Observable
class AppState {
    func todayCount(from records: [ResistRecord]) -> Int {
        let calendar = Calendar.current
        let startOfDay = calendar.startOfDay(for: Date())
        return records.filter { $0.createdAt >= startOfDay }.count
    }

    func thisWeekCount(from records: [ResistRecord]) -> Int {
        let calendar = Calendar.current
        guard let startOfWeek = calendar.dateInterval(of: .weekOfYear, for: Date())?.start else { return 0 }
        return records.filter { $0.createdAt >= startOfWeek }.count
    }

    func thisMonthCount(from records: [ResistRecord]) -> Int {
        let calendar = Calendar.current
        guard let startOfMonth = calendar.dateInterval(of: .month, for: Date())?.start else { return 0 }
        return records.filter { $0.createdAt >= startOfMonth }.count
    }

    func thisYearCount(from records: [ResistRecord]) -> Int {
        let calendar = Calendar.current
        guard let startOfYear = calendar.dateInterval(of: .year, for: Date())?.start else { return 0 }
        return records.filter { $0.createdAt >= startOfYear }.count
    }
    
    func streakDays(from records: [ResistRecord]) -> Int {
        guard !records.isEmpty else { return 0 }
        
        let calendar = Calendar.current
        let sortedDates = records.map { calendar.startOfDay(for: $0.createdAt) }
            .sorted(by: >)
        
        var streak = 0
        var checkDate = calendar.startOfDay(for: Date())
        
        if !sortedDates.contains(checkDate) {
            checkDate = calendar.date(byAdding: .day, value: -1, to: checkDate)!
        }
        
        var dateSet = Set(sortedDates)
        while dateSet.contains(checkDate) {
            streak += 1
            checkDate = calendar.date(byAdding: .day, value: -1, to: checkDate)!
        }
        
        return streak
    }

    func bestStreak(from records: [ResistRecord]) -> Int {
        guard !records.isEmpty else { return 0 }
        let calendar = Calendar.current
        let dateSet = Set(records.map { calendar.startOfDay(for: $0.createdAt) })
        let sortedDates = dateSet.sorted()
        var best = 1
        var current = 1
        for i in 1..<sortedDates.count {
            let diff = calendar.dateComponents([.day], from: sortedDates[i - 1], to: sortedDates[i]).day ?? 0
            if diff == 1 {
                current += 1
                best = max(best, current)
            } else {
                current = 1
            }
        }
        return best
    }
    
    func totalCount(from records: [ResistRecord]) -> Int {
        records.count
    }
    
    func totalSavedAmount(from records: [ResistRecord]) -> Double {
        records.compactMap { $0.amount }.reduce(0, +)
    }

    // MARK: - 节省金额详细统计

    /// 本月节省金额
    func thisMonthSavedAmount(from records: [ResistRecord]) -> Double {
        let calendar = Calendar.current
        guard let startOfMonth = calendar.dateInterval(of: .month, for: Date())?.start else { return 0 }
        return records.filter { $0.createdAt >= startOfMonth }.compactMap { $0.amount }.reduce(0, +)
    }

    /// 今年节省金额
    func thisYearSavedAmount(from records: [ResistRecord]) -> Double {
        let calendar = Calendar.current
        guard let startOfYear = calendar.dateInterval(of: .year, for: Date())?.start else { return 0 }
        return records.filter { $0.createdAt >= startOfYear }.compactMap { $0.amount }.reduce(0, +)
    }

    /// 本周节省金额
    func thisWeekSavedAmount(from records: [ResistRecord]) -> Double {
        let calendar = Calendar.current
        guard let startOfWeek = calendar.dateInterval(of: .weekOfYear, for: Date())?.start else { return 0 }
        return records.filter { $0.createdAt >= startOfWeek }.compactMap { $0.amount }.reduce(0, +)
    }

    /// 今天节省金额
    func todaySavedAmount(from records: [ResistRecord]) -> Double {
        let calendar = Calendar.current
        let startOfDay = calendar.startOfDay(for: Date())
        return records.filter { $0.createdAt >= startOfDay }.compactMap { $0.amount }.reduce(0, +)
    }

    /// 按分类的节省金额统计
    func savedAmountByCategory(from records: [ResistRecord]) -> [(emoji: String, name: String, categoryID: String, amount: Double, count: Int)] {
        let withAmount = records.filter { $0.amount != nil }
        let grouped = Dictionary(grouping: withAmount) { $0.effectiveCategoryID }
        return grouped.map { catID, items in
            let emoji = items.first?.categoryEmoji ?? ""
            let fallback = items.first?.category ?? catID
            let name = localizedCategoryName(categoryID: catID, fallback: fallback)
            let total = items.compactMap { $0.amount }.reduce(0, +)
            return (emoji, name, catID, total, items.count)
        }.sorted { $0.amount > $1.amount }
    }

    /// 月度节省金额趋势（近6个月）
    func monthlySavedAmountTrend(from records: [ResistRecord]) -> [(label: String, amount: Double)] {
        let calendar = Calendar.current
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM"

        let grouped = Dictionary(grouping: records) { formatter.string(from: $0.createdAt) }

        var data: [(String, Double)] = []
        for offset in stride(from: -5, through: 0, by: 1) {
            if let date = calendar.date(byAdding: .month, value: offset, to: Date()) {
                let key = formatter.string(from: date)
                let amount = grouped[key]?.compactMap { $0.amount }.reduce(0, +) ?? 0
                let displayFormatter = DateFormatter()
                displayFormatter.locale = Locale.current
                displayFormatter.setLocalizedDateFormatFromTemplate("MMM")
                let label = displayFormatter.string(from: date)
                data.append((label, amount))
            }
        }
        return data
    }

    /// 节省金额最多的一天
    func bestSavedDay(from records: [ResistRecord]) -> (date: Date, amount: Double)? {
        let calendar = Calendar.current
        let withAmount = records.filter { $0.amount != nil }
        guard !withAmount.isEmpty else { return nil }
        let grouped = Dictionary(grouping: withAmount) { calendar.startOfDay(for: $0.createdAt) }
        let best = grouped.map { (date, items) in
            (date, items.compactMap { $0.amount }.reduce(0, +))
        }.max { $0.1 < $1.1 }
        guard let best, best.1 > 0 else { return nil }
        return best
    }

    /// 平均每次节省金额（仅含填写了金额的记录）
    func averageSavedAmount(from records: [ResistRecord]) -> Double {
        let amounts = records.compactMap { $0.amount }
        guard !amounts.isEmpty else { return 0 }
        return amounts.reduce(0, +) / Double(amounts.count)
    }

    /// 含有金额的记录数
    func recordsWithAmountCount(from records: [ResistRecord]) -> Int {
        records.filter { $0.amount != nil }.count
    }

    // MARK: - Widget 数据同步

    /// 将关键统计数据写入 App Group 共享容器，供 Widget Extension 读取
    func syncWidgetData(from records: [ResistRecord]) {
        let stats = WidgetDataStore.Stats(
            todayCount:     todayCount(from: records),
            totalCount:     totalCount(from: records),
            streakDays:     streakDays(from: records),
            totalSaved:     totalSavedAmount(from: records),
            currencySymbol: SupportedCurrency.systemSymbol
        )
        WidgetDataStore.save(stats: stats)
        WidgetCenter.shared.reloadTimelines(ofKind: "HoldItWidget")
    }
}

File: hold_it_flutter/lib/viewmodels/app_state.dart
import 'package:flutter/material.dart';
import '../models/models.dart';
import '../database/database_helper.dart';
import 'package:sqflite/sqflite.dart';

class AppState extends ChangeNotifier {
  List<ResistRecord> _records = [];
  List<CustomCategory> _customCategories = [];
  List<Reward> _rewards = [];
  List<RewardCoinRecord> _coinRecords = [];

  List<ResistRecord> get records => _records;
  List<CustomCategory> get customCategories => _customCategories;
  List<Reward> get rewards => _rewards;
  List<RewardCoinRecord> get coinRecords => _coinRecords;

  AppState() {
    loadData();
  }

  Future<void> loadData() async {
    final db = await DatabaseHelper.instance.database;
    final recordMaps = await db.query('resist_records', orderBy: 'created_at DESC');
    final categoryMaps = await db.query('custom_categories', orderBy: 'created_at ASC');
    final rewardMaps = await db.query('rewards');
    final coinMaps = await db.query('reward_coin_records', orderBy: 'created_at DESC');

    _records = recordMaps.map((m) => ResistRecord.fromMap(m)).toList();
    _customCategories = categoryMaps.map((m) => CustomCategory.fromMap(m)).toList();
    _rewards = rewardMaps.map((m) => Reward.fromMap(m)).toList();
    _coinRecords = coinMaps.map((m) => RewardCoinRecord.fromMap(m)).toList();
    notifyListeners();
  }

  Future<void> addRecord(ResistRecord record) async {
    final db = await DatabaseHelper.instance.database;
    await db.insert('resist_records', record.toMap());
    await loadData();
  }

  Future<void> addCustomCategory(CustomCategory category) async {
    final db = await DatabaseHelper.instance.database;
    await db.insert('custom_categories', category.toMap());
    await loadData();
  }

  Future<void> deleteCustomCategory(String id) async {
    final db = await DatabaseHelper.instance.database;
    await db.delete('custom_categories', where: 'id = ?', whereArgs: [id]);
    await loadData();
  }

  Future<void> addReward(Reward reward) async {
    final db = await DatabaseHelper.instance.database;
    await db.insert('rewards', reward.toMap());
    await loadData();
  }

  Future<void> updateReward(Reward reward) async {
    final db = await DatabaseHelper.instance.database;
    await db.update('rewards', reward.toMap(), where: 'id = ?', whereArgs: [reward.id]);
    await loadData();
  }

  Future<void> deleteReward(String id) async {
    final db = await DatabaseHelper.instance.database;
    await db.delete('rewards', where: 'id = ?', whereArgs: [id]);
    await db.delete('reward_coin_records', where: 'reward_id = ?', whereArgs: [id]);
    await loadData();
  }

  Future<void> addCoinRecord(RewardCoinRecord coinRecord) async {
    final db = await DatabaseHelper.instance.database;
    await db.insert('reward_coin_records', coinRecord.toMap());
    await loadData();
  }

  Future<void> clearAllData() async {
    final db = await DatabaseHelper.instance.database;
    await db.delete('resist_records');
    await db.delete('custom_categories');
    await db.delete('rewards');
    await db.delete('reward_coin_records');
    await loadData();
  }

  // MARK: - Statistics

  int todayCount() {
    final startOfDay = DateTime.now().copyWith(hour: 0, minute: 0, second: 0, millisecond: 0);
    return _records.where((r) => r.createdAt.isAfter(startOfDay) || r.createdAt.isAtSameMomentAs(startOfDay)).length;
  }

  int thisWeekCount() {
    final now = DateTime.now();
    final startOfWeek = now.subtract(Duration(days: now.weekday - 1)).copyWith(hour: 0, minute: 0, second: 0, millisecond: 0);
    return _records.where((r) => r.createdAt.isAfter(startOfWeek) || r.createdAt.isAtSameMomentAs(startOfWeek)).length;
  }

  int thisMonthCount() {
    final now = DateTime.now();
    final startOfMonth = DateTime(now.year, now.month, 1);
    return _records.where((r) => r.createdAt.isAfter(startOfMonth) || r.createdAt.isAtSameMomentAs(startOfMonth)).length;
  }

  int thisYearCount() {
    final now = DateTime.now();
    final startOfYear = DateTime(now.year, 1, 1);
    return _records.where((r) => r.createdAt.isAfter(startOfYear) || r.createdAt.isAtSameMomentAs(startOfYear)).length;
  }

  int streakDays() {
    if (_records.isEmpty) return 0;
    final sortedDates = _records.map((r) => DateTime(r.createdAt.year, r.createdAt.month, r.createdAt.day)).toSet().toList();
    sortedDates.sort((a, b) => b.compareTo(a));

    var streak = 0;
    var checkDate = DateTime.now().copyWith(hour: 0, minute: 0, second: 0, millisecond: 0);
    if (!sortedDates.any((d) => d.isAtSameMomentAs(checkDate))) {
      checkDate = checkDate.subtract(const Duration(days: 1));
    }

    final dateSet = sortedDates.toSet();
    while (dateSet.any((d) => d.isAtSameMomentAs(checkDate))) {
      streak++;
      checkDate = checkDate.subtract(const Duration(days: 1));
    }
    return streak;
  }

  int bestStreak() {
    if (_records.isEmpty) return 0;
    final dateSet = _records.map((r) => DateTime(r.createdAt.year, r.createdAt.month, r.createdAt.day)).toSet().toList();
    dateSet.sort();
    var best = 1;
    var current = 1;
    for (int i = 1; i < dateSet.length; i++) {
      final diff = dateSet[i].difference(dateSet[i - 1]).inDays;
      if (diff == 1) {
        current++;
        if (current > best) best = current;
      } else {
        current = 1;
      }
    }
    return best;
  }

  int totalCount() => _records.length;

  double totalSavedAmount() => _records.where((r) => r.amount != null).fold(0.0, (sum, r) => sum + (r.amount ?? 0));

  double thisMonthSavedAmount() {
    final now = DateTime.now();
    final startOfMonth = DateTime(now.year, now.month, 1);
    return _records
        .where((r) => r.createdAt.isAfter(startOfMonth) || r.createdAt.isAtSameMomentAs(startOfMonth))
        .where((r) => r.amount != null)
        .fold(0.0, (sum, r) => sum + (r.amount ?? 0));
  }

  double thisYearSavedAmount() {
    final now = DateTime.now();
    final startOfYear = DateTime(now.year, 1, 1);
    return _records
        .where((r) => r.createdAt.isAfter(startOfYear) || r.createdAt.isAtSameMomentAs(startOfYear))
        .where((r) => r.amount != null)
        .fold(0.0, (sum, r) => sum + (r.amount ?? 0));
  }

  double thisWeekSavedAmount() {
    final now = DateTime.now();
    final startOfWeek = now.subtract(Duration(days: now.weekday - 1)).copyWith(hour: 0, minute: 0, second: 0, millisecond: 0);
    return _records
        .where((r) => r.createdAt.isAfter(startOfWeek) || r.createdAt.isAtSameMomentAs(startOfWeek))
        .where((r) => r.amount != null)
        .fold(0.0, (sum, r) => sum + (r.amount ?? 0));
  }

  double todaySavedAmount() {
    final startOfDay = DateTime.now().copyWith(hour: 0, minute: 0, second: 0, millisecond: 0);
    return _records
        .where((r) => r.createdAt.isAfter(startOfDay) || r.createdAt.isAtSameMomentAs(startOfDay))
        .where((r) => r.amount != null)
        .fold(0.0, (sum, r) => sum + (r.amount ?? 0));
  }

  List<Map<String, dynamic>> savedAmountByCategory() {
    final withAmount = _records.where((r) => r.amount != null).toList();
    final grouped = <String, List<ResistRecord>>{};
    for (final r in withAmount) {
      final key = r.effectiveCategoryID;
      grouped.putIfAbsent(key, () => []).add(r);
    }
    return grouped.entries.map((e) {
      final items = e.value;
      final emoji = items.first.categoryEmoji;
      final name = items.first.category;
      final total = items.where((r) => r.amount != null).fold(0.0, (sum, r) => sum + (r.amount ?? 0));
      return {
        'emoji': emoji,
        'name': name,
        'categoryID': e.key,
        'amount': total,
        'count': items.length,
      };
    }).toList()
      ..sort((a, b) => (b['amount'] as double).compareTo(a['amount'] as double));
  }

  List<Map<String, dynamic>> monthlySavedAmountTrend() {
    final now = DateTime.now();
    final result = <Map<String, dynamic>>[];
    for (int offset = -5; offset <= 0; offset++) {
      final date = DateTime(now.year, now.month + offset, 1);
      final key = '${date.year}-${date.month.toString().padLeft(2, '0')}';
      final amount = _records
          .where((r) {
            final rKey = '${r.createdAt.year}-${r.createdAt.month.toString().padLeft(2, '0')}';
            return rKey == key && r.amount != null;
          })
          .fold(0.0, (sum, r) => sum + (r.amount ?? 0));
      final label = _monthLabel(date);
      result.add({'label': label, 'amount': amount});
    }
    return result;
  }

  String _monthLabel(DateTime date) {
    final labels = ['1月', '2月', '3月', '4月', '5月', '6月', '7月', '8月', '9月', '10月', '11月', '12月'];
    return labels[date.month - 1];
  }

  Map<String, dynamic>? bestSavedDay() {
    final withAmount = _records.where((r) => r.amount != null).toList();
    if (withAmount.isEmpty) return null;
    final grouped = <DateTime, double>{};
    for (final r in withAmount) {
      final day = DateTime(r.createdAt.year, r.createdAt.month, r.createdAt.day);
      grouped[day] = (grouped[day] ?? 0) + (r.amount ?? 0);
    }
    final best = grouped.entries.reduce((a, b) => a.value > b.value ? a : b);
    if (best.value <= 0) return null;
    return {'date': best.key, 'amount': best.value};
  }

  double averageSavedAmount() {
    final amounts = _records.where((r) => r.amount != null).map((r) => r.amount!).toList();
    if (amounts.isEmpty) return 0;
    return amounts.reduce((a, b) => a + b) / amounts.length;
  }

  int recordsWithAmountCount() => _records.where((r) => r.amount != null).length;
}
